test/Chainweb/Test/Pact/ForkTest.hs

```diff
@@ -0,0 +1,342 @@
+    withPayloadDb $ \pdb ->
+    testProperty "prop_forkValidates" (prop_forkValidates pdb bhdb cid _genBlock)
+  where
+    _genBlock = genesisBlockHeader testVer cid
+    cid = someChainId testVer
+
+scheduledTest :: ScheduledTest
+scheduledTest = ScheduledTest "Pact checkpointer forking test" test
+
+testVer :: ChainwebVersion
+testVer = FastTimedCPM petersonChainGraph
+
+-- | Property: Fork requiring checkpointer rewind validates properly
+prop_forkValidates
+    :: IO (PayloadDb HashMapCas)
+    -> (IO BlockHeaderDb)
+    -> ChainId
+    -> BlockHeader
+    -> Property
+prop_forkValidates pdb bhdb cid genBlock =
+    again $ ioProperty $ do
+        (trunk, left, right) <- generate genForkLengths
+        mVar <- newMVar (0 :: Int)
+        withPactProp testVer Warn pdb bhdb (testMemPoolAccess cid mVar) (return Nothing) $ \reqQ -> do
+            db <- bhdb
+            putStrLn $ "Testing fork lengths:"
+                        ++ " trunk: " ++ show trunk
+                        ++ ", left: " ++ show left
+                        ++ ", right: " ++ show right
+            expected <- expectedForkProd (trunk, left, right)
+            if expected >= maxBalance
+              then do -- this shouldn't happen...
+                putStrLn "Max account balance would be exceeded, letting this test pass"
+                return $ property Discard
+              else do
+                (_nbTrunkRes, _vbTrunkRes, parentFromTrunk) <-
+                    runBlocks db genBlock (trunk - 2) reqQ bhdb -- '-2' for genesis block and 0-based
+                (_nbLeftRes, _vbLeftRes, _parentFromLeft) <- liftIO $
+                    runBlocks db parentFromTrunk (left - 1) reqQ bhdb
+                (nbRightRes, vbRightRes, _parentFromRight) <- liftIO $
+                    runBlocks db parentFromTrunk (right - 1) reqQ bhdb
+                return $ property (nbRightRes == vbRightRes)
+
+forkTestQueueSize :: Natural
+forkTestQueueSize = 1024
+
+withPactProp
+    :: ChainwebVersion
+    -> LogLevel
+    -> IO (PayloadDb HashMapCas)
+    -> IO BlockHeaderDb
+    -> MemPoolAccess
+    -> IO (Maybe FilePath)
+    -> (PactQueue -> IO Property)
+    -> IO Property
+withPactProp version logLevel iopdb iobhdb mempool iodir f =
+    bracket startPact stopPact $ \(_x, q) -> f q
+  where
+    startPact :: IO (Async (), TBQueue RequestMsg)
+    startPact = do
+        reqQ <- atomically $ newTBQueue forkTestQueueSize
+        pdb <- iopdb
+        bhdb <- iobhdb
+        dir <- iodir
+        a <- async $ PS.initPactService version cid logger reqQ mempool
+                         bhdb pdb dir Nothing True -- last Param True means reset checkpointer db
+        return (a, reqQ)
+
+    stopPact :: (Async a, TBQueue a2) -> IO ()
+    stopPact (a, _) = cancel a
+
+    logger = genericLogger logLevel T.putStrLn
+    cid = someChainId version
+
+genForkLengths :: Gen (Int, Int, Int)
+genForkLengths = do
+    let maxTotalLen = 12
```

Comment:
These forks, and the total branch depth, aren't very deep. I don't know whether that matters for us here but e.g. for the new checkpointer, this mightn't even be deep enough to trigger the compaction logic

Reply:
I had to control the length here to avoid hitting the max rewind of 8 on newBlock -- is that not going to be an issue with the new version?

Reply:
As per discussion, I'm going to make the depth configurable in another PR